From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sven Strickroth <email@cs-ware.de>
Date: Sun, 3 Mar 2013 19:06:44 +0100
Subject: Reset decoration cache

This is needed for RevisionGraph to include new ref and excluded deleted ones.

Related to issue #1663.

Signed-off-by: Sven Strickroth <email@cs-ware.de>
---
 decorate.c | 19 +++++++++++++++++++
 decorate.h |  1 +
 log-tree.c |  9 +++++++++
 3 files changed, 29 insertions(+)

diff --git a/decorate.c b/decorate.c
index 9f24925263..67e71d51fd 100644
--- a/decorate.c
+++ b/decorate.c
@@ -32,6 +32,25 @@ static void *insert_decoration(struct decoration *n, const struct object *base,
 	return NULL;
 }
 
+void free_decoration(struct decoration *n)
+{
+	struct decoration_entry *old_entries;
+
+	if (!n || !n->entries)
+		return;
+
+	old_entries = n->entries;
+
+	for (unsigned int i = 0; i < n->size; ++i) {
+		if (old_entries[i].base && old_entries[i].decoration)
+			free(old_entries[i].decoration);
+	}
+
+	FREE_AND_NULL(n->entries);
+	n->size = 0;
+	n->nr = 0;
+}
+
 static void grow_decoration(struct decoration *n)
 {
 	unsigned int i;
diff --git a/decorate.h b/decorate.h
index 08af658d34..fa3c359419 100644
--- a/decorate.h
+++ b/decorate.h
@@ -57,6 +57,7 @@ void *add_decoration(struct decoration *n, const struct object *obj, void *decor
  * association, this function returns NULL.
  */
 void *lookup_decoration(struct decoration *n, const struct object *obj);
+extern void free_decoration(struct decoration *n);
 
 /*
  * Clear all decoration entries, releasing any memory used by the structure.
diff --git a/log-tree.c b/log-tree.c
index 9de6548445..4590384812 100644
--- a/log-tree.c
+++ b/log-tree.c
@@ -217,6 +217,15 @@ static int add_graft_decoration(const struct commit_graft *graft,
 	return 0;
 }
 
+void clear_ref_decorations(void)
+{
+	if (decoration_loaded)
+	{
+		free_decoration(&name_decoration);
+		decoration_loaded = 0;
+	}
+}
+
 void load_ref_decorations(struct decoration_filter *filter, int flags)
 {
 	if (!decoration_loaded) {
